### DISCUSSION TITLE
Is there a method that runs only once server-side on first page request?

### DISCUSSION THREAD

**Discussion by ahmadsdn** [vercel/next.js#18210] [Help] [answered]
I'm new to next.js and I'm looking for a method (similar to `getServerSideProps`) that runs just once on the server. `getServerSideProps` runs on first page request but also runs on client side navigation (with an ajax request).

The best solution I could find is to use `getInitialProps` and by checking `ctx.req` to find out if it is being executed on client retrun an empty object.

But I think it is not ideal, I may want to make some API calls that don't want the code to be exposed to the client.
Any suggestions?

**Comment by jamesmosier** [comment, marked as answer]
There is no mechanism within Next to do this, other than using `getInitialProps` as you mentioned. 

If you do this then your page will not have the data it needs when you navigate to it via `next/link`. For example if you are on `/page1` and then click a `next/link` to `/page2`, `page2` will not know what data to have and will behave differently if `getServerSideProps` logic is not called. That is why `getServerSideProps` runs on the server and then makes a `fetch` call to get the data on the client. 


> I may want to make some API calls that don't want the code to be exposed to the client

There is no difference between this and rendering on the server - the data is still available to the client. There are no security issues because the client only receives the data you send it. If you don't want the client to receive some data then exclude it from being returned.


---

If you really want to do this and use `getServerSideProps` you could look at `ctx.req.url` which is different depending on server side rendering and when it is called as an API on the client.

**Reply by ahmadsdn** [reply]
Thanks for your input.
I have a context provider in `_app.js` wrapping the whole application. I want to get user data in `_app.js` and make it available to entire application. I just need to get it once on the server the when the page loads it is available and I don't need to get is on the client.
And it also persists when the user navigates on the client via `next/link` so I do not have to refetch the data when the URL changes.

I believe the code written inside `getServerSideProps` is not included in client js files, so it is safe to use say secret API keys in this method. But the code inside `getInitialProps` is also shipped within js files to client. So no secrets should be included in this function.

It would be nice if there were such a method that would run only once on server (for initial page render server-side).

**Reply by jamesmosier** [reply]
Oh sorry, I didn't realize you were talking about globally in your `_app.js`! Yeah as mentioned, there is no way to do this currently for server only code 😞

**Reply by nimeshvaghasiya** [reply]
@jamesmosier is there any way now to globally share api data (called only once per user login) from _app.tsx? I'm using next v12.2.2.

**Reply by aldabil21** [reply]
> @jamesmosier is there any way now to globally share api data (called only once per user login) from _app.tsx? I'm using next v12.2.2.

If your case:
1. Does not accept to get the data client-side, and must be fetched once on server before the first render.
2. Does not accept to use `getInitialProps` in `_app`.

Then, you would have to wait for #37136

**Reply by SamirMishra27** [reply]
Since this is out now

@aldabil21 So I will just have to put the logic in `app/layout.tsx` file right? I'm using Next's App router and need to run a function only once first when the user visits the page. (I'm on 13.4)

**Reply by aldabil21** [reply]
Yes the root layout

**Comment by patjakubik** [comment]
Hey @asajadian - what did you end doing? I'm struggling with the exactly same problem.

**Reply by ahmadsdn** [reply]
Hey @patryk-smc , sorry I haven't been around for a long time.
You might have found a better solution by now, but as I had mentioned in my question the best solution for me was using `getInitialProps`.

I used it in _app.js like so:

```
App.getInitialProps = async (context) => {
  const { router, ctx } = context;
  const isServer = !!ctx.req;

  if(isServer){
  // do things you want to do on server (initial render)
  }
  else{
  // things you want in client side
  }
};
```

**Reply by denu5** [reply]
hi @asajadian, as far as i can see the check for ctx.req doesnt work since its still true for client and server...?

as @jamesmosier suggests that check for different ctx.req.url works, but this solutions feels somehow wrong to me:  ```ctx.req?.url?.indexOf('/_next/data/') === -1)```

---

-> my goal is to have an initial call to retrieve a global page navigation tree and then persist the state after page navigations (using ssr) as its sold here: https://nextjs.org/docs/advanced-features/custom-app

I'm surprised that I can't find a working example how to properly achieve **"a method that runs only once server-side on first page request"** or is this maybe generally not the right thinking in the next.js world ? 

The Q/A here is the closest that comes to a working solution, but its missing the "call only once" part : https://stackoverflow.com/questions/60899880/next-js-reduce-data-fetching-and-share-data-between-pages


...searching for some more answers I can see that people get often pointed to using the redux nextjs wrapper, but would like to avoid getting into that complex setups already.

https://stackoverflow.com/questions/63180045/pass-data-from-page-to-another
https://stackoverflow.com/questions/57013202/managing-users-global-state-on-next-js-application
https://stackoverflow.com/questions/61450721/how-to-persist-state-across-pages-using-nextjs-and-usecontext-hook

**Reply by denu5** [reply]
my hacky idea looks something like this: 
```
function MyApp({ Component, pageProps }: AppProps) {
    const [pageNav, setPageNav] = useState<any>();

    useEffect(() => {
        setPageNav(pageProps.pageNav);
    }, []);

    return (
                <Layout pageNav={pageNav}>
                    <Component {...pageProps} />
                </Layout>
    );
}

MyApp.getInitialProps = async ({ router, ctx, Component }: AppContext) => {
    const pageProps: any = {
        ...(Component.getInitialProps ? await Component.getInitialProps(ctx) : {}),
    };

    if (ctx.req?.url?.indexOf('/_next/data/') === -1) {    //  /_next/data/ are the client page calls to the server
        console.log('is server');
        pageProps.pageNav = await getBaseContentLayout(router.locale!);
    } else {
        console.log('is client');
    }
    return { pageProps };
};
export default MyApp;
```

**Reply by ahmadsdn** [reply]
hi @denu5 `ctx.req` works, try `console.log(!!ctx.req)` in `getInitialProps` you'll see it is true for first render (in your dev terminal) and it's false when you navigate client side (browser console).

I'm using this for authentication. I do the check in `_app.js` if it is on the server I make an api call to fetch user data but if it is on the client I check ` window.__NEXT_DATA__.props` to determine if user is logged in or not.

Keep in mind though that the `context` object seems to have different properties if it is used in `_app.js` or in a page. If it is in `_app.js` you need to check `context.ctx.req` in pages it is `context.req`.

**Reply by denu5** [reply]
which version of next.js are you using? 

in my case I ```console.log(ctx.req.url);```

on the first call I see the "normal url" (server)
...and on subsequent navigations you can see the /_next/data......json urls

... so having said the above, ctx.req is always true in my case

**Reply by ahmadsdn** [reply]
I'm using version `9.5.2`.
On subsequent navigations `ctx.req` is `undefined` in my case.

**Reply by denu5** [reply]
Ah okey, I'm using the latest 10.0.5 from 1d ago 😬

**Reply by patjakubik** [reply]
I solved it with something different. I created a provider which wraps the whole app, and on the first page request - show the loader, and make API call from the client - to the server. Furthermore, I like that approach because it doesn't use getInitialProps, making it framework-agnostic, and moreover, Automatic Static Optimization is turned on 😄 

The only disadvantage is that getServersideProps cannot be used if you need data that is only obtained at the stage of first page request.

**Reply by denu5** [reply]
@patryk-smc going that way was also one idea but then I was thinking that having the "navigation-structure" in ssr for seo could be better...

**Reply by stefvhuynh** [reply]
@patryk-smc that solution will send an unpopulated HTML though. it won't work if you want SEO to pick up data in your HTML.

**Reply by Yankovsky** [reply]
It seems like @denu5 hacky solution is the only one that works.
I've also noticed, that App#getInitialProps is called on server side if the page you are making client-navigation to has getServerSideProps, but at the same time App#getInitialProps is called on client side if the next page has no getServerSideProps.

**Comment by Yankovsky** [comment]
Is there a way? I can provide some examples, where such thing would be nice to have:
- I want to determine if user has a mobile device by parsing User Agent string; I want to do that on server and only once per session (before page reload)
- I want to get user's a/b group; I want to get it once from server side and then cache for user session duration.

**Comment by Yankovsky** [comment]
So guys, I've implemented something like this:
```ts
function MyApp({ Component, pageProps }: AppProps) {
  // Save these guys in refs so they will stay the same no matter what.
  const ab = useRef(pageProps.ab)
  const isLaptop = useRef(pageProps.isLaptop)

  return (
    <ABProvider ab={ab.current}>
      <IsLaptopProvider isLaptop={isLaptop.current}>
        <Component {...pageProps} />
      </IsLaptopProvider>
    </ABProvider>
  )
}

MyApp.getInitialProps = async (appContext: AppContext) => {
  const appProps = await App.getInitialProps(appContext)

  // Webpack will not include this in client bundle
  if (typeof window === 'undefined') {
    // Hacky check from mister @denu5
    // Ensures that the expression under this condition will be executed
    // only once on first page render.
    // If the url starts with /_next/data, that means that we have
    // a client-side navigation situation requiring some additional data
    // from getServerSideProps or getInitialProps (not sure about this one).
    if (!appContext.ctx.req?.url?.startsWith('/_next/data')) {
      // Calculate something on server side
      appProps.pageProps.isLaptop = isLaptop(appContext.ctx.req)
      appProps.pageProps.ab = abMiddleware(appContext.ctx)
    }
  }

  return { ...appProps }
}

export default MyApp
```

**Reply by dmwin72015** [reply]
it works ,but not look so good!

**Reply by Yankovsky** [reply]
agreed 😄

**Reply by denu5** [reply]
> agreed 😄 

with the new middleware concept in  v12 this thing might be worth refactoring depending on what you need

https://github.com/vercel/examples/tree/main/edge-functions/ab-testing-simple

**Reply by Faisal-Manzer** [reply]
It works great.

**Reply by behzadev** [reply]
@denu5 how using a middleware could resolve the OP request?

**Comment by theodorebear** [comment]
Hey guys, I am looking for something similar, but within next-redux-wrapper.
Maybe this is along the same lines as what you need: https://github.com/kirill-konshin/next-redux-wrapper/issues/436

**Comment by tannguyencse19** [comment]
My solution for `firebase-admin`, I only need initialized one time.

```jsx
export const getStaticProps: GetStaticProps = async (
  context
) => {
  if (getApps().length === 0) initializeApp(ServerConfig); // <-- Key is here!

  getDatabase()
    .ref("/destination")
    .on(
      "value",
      (snapshot) => {
        console.log(snapshot.val());
      },
      (errorObject) => {
        console.log("The read failed: " + errorObject.name);
      }
    );

  return {
    props: {
     
      },
    },
  };
};
```

Not sure if multiple reinitialize cause any damage?

---
Furthermore, you can refactoring like this

```jsx
export const DB = () => {
  if (getApps().length === 0) initializeApp(ServerConfig);

  return getDatabase();
};
```

**Comment by PredokMiF** [comment]
If we will use `getInitialProps` then compiler will not remove code and packages that needed on serverside. In case above, with `if` in `getInitialProps` we cant use `pg` fore example, cause in brouser it is useless, but it wil be placed to froduction frontend bundle.

**Comment by skywalker282** [comment]
Like Vercel team doesn't provide any support yet, we should use separate service for everythink we need to link on the server startup. So that we can lunch theme just after having launched our next app. I did it for a cron job and everything is working good now. Happy debugging
